---
title: "RAREsim Vignette V1"
author: "Megan Null"
date: "5/25/2020"
output:
  rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{RAREsim Vignette V1}
  \usepackage[UTF-8]{inputenc}  
---
This vigette describes how to use the RAREsim R package to simulate rare variant genetic data. 

The RAREsim R package currently provides functions:  

* Expected_variants  
* Fit_AFS  
* Fit_total_variants
* Total_variants  
* Prune

Here we will walk through an example simulation of RAREsim using one cM block on chromosome 19 and the African ancestry group from gnomAD.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install the package

```{r }
#install.packages('RAREsim')
library(RAREsim)
```

RAREsim has three main steps: (1) simulating genetic data with an abundance of rare variants, (2) estimating the expected number of variants in MAC bins, and (3) pruning the rare variants to match the estimated number of variants in each MAC bin.

Simulating genetic data is done using HAPGEN2 with information at all bp, including those bp that are monomorphic in the input simulation data. Within HAPGEN2, the vast majority of simulated rare variants arise from a mutation from an all reference bp within the input simulated data. Thus, we add information at all bp, including bp with all reference alleles, to allow for an abundance of rare variants.

In order to prune the simulated data to approximate real data, RAREsim estimates the number of variants within MAC bins. This is either done by using default parametes, or by fitting target data. Here we will demonstrate fitting target data.

## Fitting the Total Number of Variants function

The first function is the Total Number of Variants function. For a given region, the Total Number of Variants function estimates the number of variants per Kb. This is done by estimating $\phi$ and $\omega$ to minimize the function $f_{Var}(n)=\phi n^\omega$.

The Total Number of Variants target data consists of various sample sizes and the observed number of variants per Kb in the region of interest. Ancestry specific data is advised. 

RAREsim has example data within the R package, with the African Total Number of Variants target data shown here.

The first column is the sample size in number of individuals and the second column is the observed number of variants in the region of interest.

```{r }
# load the data
data("tot_num_var_afr")
print(tot_num_var_afr)
```

XXX figure out how to get rid of the weird row names

The target data is used to estimate $\phi$ and $\omega$ within a least squares loss function, optimizing using sequential quadratic programming. It is implemented with the *Fit_total_variants* function.

```{r }
Fit_total_variants(tot_num_var_afr)
```

XXX make the output labeled with phi and omega

The output of the Fit_total_variants function are the parameters $\phi$ and $\omega$, respectively. The estimated parameters can then be used to estimate the number of variants per Kb within the region of interest, given the number of individuals to be simulated $N_{sim}$. 

For example, to simulate the sample size observed in gnomAD for the African ancestry group ($N_{sim}=8128), we calculate $f_{Var}(N_{sim})=\phi N_{sim}^\omega$. This can be done with the *Total_number_variants* function. A value for phi, omega, and the sample size is required.


```{r }
Total_number_variants(phi = 0.1638108, omega = 0.6248848, n = 8128)
```

XXXX in the perfect world, this would show the fit of the function for the one block

The number of variants per Kb was estimated using parameters fit with the target data. However, RAREsim also provides ancestry specific default parameters that can be used to estimate the number of variants per Kb.

XXX add default parameters

```{r }
#Total_number_variants(0.1638108, 0.6248848, 8128) #with default
```

Now this is the number of variants per Kb. The cM block we are simulating has a total of 19,029 bp. Thus, to get the total expected number of variants in the region, we multiple the expected number of variants per Kb (Total_number_variants) by 19.029.

XXX talk about capture/recapture?

```{r }
19.029*Total_number_variants(phi = 0.1638108, omega = 0.6248848, n = 8128)
```

At this point, we know how many variants we expect to see within the region. We now need the allele frequency spectrum (AFS) to calculate the proportion of variants at various minor allele counts.

## Fitting the Allele Frequency Spectrum function

The AFS function inputs allele frequency information and outputs estimates for $\alpha$, $\beta$, and $b$. The required AFS target data is a vector with the number of variants in each MAC bin:  
MAC=1  
MAC=2  
MAC=3-5  
MAC=6-10  
MAC=11-20  
MAC=21 - MAF = 0.5%  
MAC = 0.5% - MAF = 1%  

Below we examine in the AFS target data for the African ancestry group.

```{r }
# load the data
data("afs_afr")
print(afs_afr)
```

XXX again, remove row names

```{r }
Fit_AFS(afs_afr, N = 8128, p_rv  = 0.97)
```

XXX need to format the output with names

## Expected Number of Variants per MAC bin

Now that all the parameters have been fit, the expected number  of variants in each MAC bin can be extimated

```{r }
bin_estimates <- Expected_variants(alpha = 1.5313376 , beta = -0.3090162, b = 0.2926182, phi = 0.1638108 , omega = 0.6248848 , Ntar = 8128, Size = 19.029)
print(bin_estimates)
```


XXXX need to remove the proportion... and allow this table to be  entered. If the proportion is removed, this will be what is input into the pruning function.

XXXX right now the bins are hard coded in. Need to make them flexible to fit the rest of the functions

## Pruning Variants

In order to use RAREsim to prune simulated data, genetic data must be simulated with HAPGEN2, with all bp added to the input haplotypes. This will simulate an abudance of rare variants to allow for variant pruning.

Pruning variants required a MAC file fo the simulated data, and the expected number of variants within each MAC bin (what is produced fom the *Expected_variants* function).


The pruning function inputs the name of the gzipped haplotype file, a MAC file from the haplotypes (i.e. number of alternate alleles  at each variant), and the expected number of variants per bin data  frame, with the bins defined. 

```{r }
bin_estimates <- bin_estimates[,-c(3)] ### I need to fix the output of this function
bin_estimates
```

```{r }
data("MAC_afr")
dim(MAC_afr)
head(MAC_afr)
```

When running the pruning function the haplotypes files will be  re-written. Thus, if the original files are needed, a copy needs to be made.

```{r }
#Prune(hap_file_name = '/location/of/haplotypes/.controls.haps.gz', MAC = MAC_afr, expected = bin_estimates)
```

XXXX the propblem with the current pruning function is that it does not work on my MAC. It calls the 'sed' command, which will only run on Linux/bash (with the same sintax at least).

XXX thoughts on this? I have a few options
1. If I can figure out the sed command on a mac, then I can just have people state if they are using a mac or Linux. Windows users???
2. I can also have an option to load in the data here. It'll take a bit of time, but I know I can make it work.


## Unrelated:
Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```


