---
title: "RAREsim Vignette V1"
author: "Megan Null"
date: "5/25/2020"
output: html_document
---
This vigette describes how to use the RAREsim R package to simulate rare variant genetic data. 

The RAREsim R package currently provides functions:  

* Expected_variants  
* Fit_AFS  
* Total_variants  
* Prune

Here we will walk through an example simulation of RAREsim using one cM block on chromosome 19 and four ancestry groups: African, East Asian, Non-Finnish European, and South Asian.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install the package

```{r }
#install.packages('RAREsim')
library(RAREsim)
```

## Fitting the Total Number of Variants function

The first function is the Total Number of Variants function. For a given region, the Total Number of Variants function estimates the number of variants per Kb. This is done by estimating $\phi$ and $\omega$ to minimize the function $f_{Var}(x)=\phi x^\omega$.

## Fitting the Allele Frequency Spectrum function

The AFS function inputs allele frequency information and outputs estimates for $\alpha$, $\beta$, and $b$.

First, load in example data:

```{r }
# load the data
data("AFS_target_data_NFE")
print(afs_nfe)
```

## Pruning Variants

In order to use RAREsim to prune simulated data, genetic data must be simulated with HAPGEN2, with all bp added to the input haplotypes. This will simulate an abudance of rare variants to allow for variant pruning.

Pruning variants required a MAC file fo the simulated data, and the expected number of variants within each MAC bin.

### Expected Number of Variants with Default Parameter

The default parameters can be used to estimate the number of variants within each MAC bin to use for simulation.

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r eval=FALSE}
A <- data.frame(ref_eur  = c(pi_star = 0, pi_hat = .15),
                 obs_afr  = c(pi_star = 1, pi_hat = .85))
```
