---
title: "RAREsim Vignette V1"
author: "Megan Null"
date: "7/2/2020"
output:
  rmarkdown::word_document
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{RAREsim Vignette V1}
  \usepackage[UTF-8]{inputenc}  
---
This vigette describes how to use the RAREsim R package to simulate rare variant genetic data. 

<!-- The RAREsim R package currently provides functions:   -->

<!-- * Fit_total_variants -->
<!-- * Total_variants  -->
<!-- * Fit_AFS  -->
<!-- * Expected_variants  -->
<!-- * Prune -->

Here we will walk through an example simulation of RAREsim using one cM block on chromosome 19 and the African ancestry group from gnomAD.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install the package

```{r }
#install.packages('RAREsim')
library(RAREsim)
```

The source code for all functions within the RAREsim package can be found at <https://github.com/meganmichelle/RAREsim_package>. It currently must be downloaded through github.

RAREsim has three main steps: (1) simulating genetic data with an abundance of rare variants using [HAPGEN2](https://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html) (Su, 2011), (2) estimating the expected number of variants in MAC bins, and (3) pruning the rare variants to match the estimated number of variants in each MAC bin.

<!-- The input simulation dataset is a sample of haplotypes (e.g. 1000 Genomes haplotypes (Genomes Project, et al., 2015)) with alternative alleles coded as 1 and all reference alleles, including sequencing bases that are not polymorphic in the input simulation dataset variants, coded as 0. Within HAPGEN2, the vast majority of simulated rare variants arise from a mutation from an all reference bp within the input simulated data. Thus, we add information at all bp, including bp with all reference alleles, to allow for an abundance of rare variants. -->

<!-- Example code to include information at all bp can *soon* be found on the [RAREsim github page](https://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html). -->

An example simulation with HAPGEN2 can also *soon* be found on the [RAREsim github page](https://github.com/meganmichelle/RAREsim/blob/master/HAPGEN2_simulation_example).

In order to prune the simulated data to approximate real data, RAREsim estimates the number of variants within MAC bins. This is either done by using default parametes, modifying default parameters or by fitting target data. Additionally, if wanting to simulate data the exact sample size of observed data, the observed data can be  matched directly. Here we will demonstrate fitting target data as well as using  default parameters.

## Fitting the Variants per Kb function

The first function is the Variants per Kb function. For a given region, the Variants per Kb function estimates the number of variants per Kb. This is done by estimating $\phi$ and $\omega$ to optimize the function $f_{Var}(n)=\phi n^\omega$.

The Variants per Kb target data consists of various sample sizes and the observed number of variants per Kb in the region of interest. Ancestry specific data is advised. 

The first column is the sample size is number of individuals and the second column is the observed number of variants per  Kb in the region of interest. Here we will fit the example data for the African ancestry population, with example data available  in  the  R package.

```{r }
# load the data
data("var_per_kb_afr")
print(var_per_kb_afr)
```


The target data is used to estimate $\phi$ and $\omega$ within a least squares loss function, optimizing using sequential quadratic programming. It is implemented with the *Fit_var_per_kb* function.

```{r }
Fit_var_per_kb(var_per_kb_afr)
```

The output of the *Fit_total_variants* function are the parameters $\phi$ and $\omega$, respectively. The estimated parameters can then be used to estimate the number of variants per Kb within the region of interest, given the number of individuals to be simulated $N_{sim}$. 

For example, to simulate the sample size observed in gnomAD for the African ancestry group ($N_{sim}=8128$), we calculate $f_{Var}(N_{sim})=\phi N_{sim}^\omega$. This can be done with the *Total_number_variants* function. A value for phi, omega, and the sample size is required.


```{r }
Variants_per_Kb(phi = 0.1638108, omega = 0.6248848, n = 8128)
```

The number of variants per Kb was estimated using parameters fit with the target data. However, RAREsim also provides ancestry specific default parameters that can be used to estimate the number of variants per Kb.

Here,  the  ancestry must be specified: AFR, EAS,  NFE, or SAS.

```{r }
Variants_per_Kb(n=8128, pop = 'AFR') 
```

The cM block we are simulating has a total of 19,029 bp. Thus, to get the total expected number of variants in the region, we multiple the expected number of variants per Kb by 19.029.


```{r }
19.029*Variants_per_Kb(phi = 0.1638108, omega = 0.6248848, n = 8128)
```

At this point, we know how many total variants we expect to see within the region. We now need the allele frequency spectrum (AFS) to calculate the proportion of variants at various minor allele bins.

## Fitting the Allele Frequency Spectrum function

The AFS function inputs allele frequency information and outputs estimates for $\alpha$, $\beta$, and $b$. The required AFS target data is a vector with the number of variants in each MAC bin. The  bins used within the evaluation of RAREsim are: 

MAC=1  
MAC=2  
MAC=3-5  
MAC=6-10  
MAC=11-20  
MAC=21 - MAF = 0.5%  
MAC = 0.5% - MAF = 1%  

Below we examine in the AFS target data for the African ancestry group. The first two columns  identify the  upper and lower boundaries of each MAC bin.  The third column  specifies the observed proportion  of variants within each MAC bin in the target data.

```{r }
# load the data
data("afs_afr")
print(afs_afr)
```


```{r }
Fit_AFS(prop_df = afs_afr, N = 8128, p_rv = 0.97)
```

## Expected Number of Variants per MAC bin

Now that all the parameters have been fit, the expected number  of variants in each MAC bin can be extimated

```{r }
bin_estimates <- Expected_variants(alpha = 1.5313376 , beta = -0.3090162, b = 0.2926182, phi = 0.1638108 , omega = 0.6248848 , Ntar = 8128, Size = 19.029)
print(bin_estimates)
```

We can also extimate the number of variants per MAC using the default parameters, again  specifying  the ancestry population as AFR, EAS, NFE, or SAS.

```{r }
bin_estimates <- Expected_variants(pop='AFR' , Ntar = 8128, Size = 19.029)
print(bin_estimates)
```


XXX right now the bins are hard coded in. Need to make them flexible to fit the rest of the functions XXX

## Pruning Variants

In order to use RAREsim to prune simulated data, genetic data must be simulated with HAPGEN2, with all bp added to the input haplotypes. This will simulate an abudance of rare variants to allow for variant pruning.

Pruning variants required a MAC file fo the simulated data, and the expected number of variants within each MAC bin (what is produced fom the *Expected_variants* function).


The pruning function inputs the name of the gzipped haplotype file, a MAC file from the haplotypes (i.e. number of alternate alleles  at each variant), and the expected number of variants per bin data  frame, with the bins defined. 

Having a MAC file allows the pruning process to be really fast.  When there aren't any alleles that need to move bins.

```{r }
data("MAC_afr")
dim(MAC_afr)
head(MAC_afr)
```

When running the pruning function the haplotypes files will be  re-written. Thus, if the original files are needed, a copy needs to be made.

```{r }
 Pruning_function(hap_file_name = '/Users/megansorenson/Documents/RAREsim/Example/Block37_rep1.controls.haps.gz', MAC = MAC_afr, expected = bin_estimates, mac = 'mac')
```

The output in  R simply states tthe number  of variants that were pruned. The haplotype  files were pruned.


