<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Megan Null" />


<title>RAREsim Vignette V1</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">RAREsim Vignette V1</h1>
<h4 class="author">Megan Null</h4>
<h4 class="date">7/2/2020</h4>



<p>This vigette describes how to use the RAREsim R package to simulate rare variant genetic data.</p>
<!-- The RAREsim R package currently provides functions:   -->
<!-- * Fit_total_variants -->
<!-- * Total_variants  -->
<!-- * Fit_AFS  -->
<!-- * Expected_variants  -->
<!-- * Prune -->
<p>Here we will walk through an example simulation of RAREsim using one cM block on chromosome 19 and the African ancestry group from gnomAD.</p>
<div id="install-the-package" class="section level2">
<h2>Install the package</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">#install.packages('RAREsim')</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(RAREsim)</a></code></pre></div>
<p>The source code for all functions within the RAREsim package can be found at <a href="https://github.com/meganmichelle/RAREsim_package" class="uri">https://github.com/meganmichelle/RAREsim_package</a>. It currently must be downloaded through github.</p>
<p>RAREsim has three main steps: (1) simulating genetic data with an abundance of rare variants using <a href="https://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">HAPGEN2</a> (Su, 2011), (2) estimating the expected number of variants in MAC bins, and (3) pruning the rare variants to match the estimated number of variants in each MAC bin.</p>
<!-- The input simulation dataset is a sample of haplotypes (e.g. 1000 Genomes haplotypes (Genomes Project, et al., 2015)) with alternative alleles coded as 1 and all reference alleles, including sequencing bases that are not polymorphic in the input simulation dataset variants, coded as 0. Within HAPGEN2, the vast majority of simulated rare variants arise from a mutation from an all reference bp within the input simulated data. Thus, we add information at all bp, including bp with all reference alleles, to allow for an abundance of rare variants. -->
<!-- Example code to include information at all bp can *soon* be found on the [RAREsim github page](https://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html). -->
<p>An example simulation with HAPGEN2 can also <em>soon</em> be found on the <a href="https://github.com/meganmichelle/RAREsim/blob/master/HAPGEN2_simulation_example">RAREsim github page</a>.</p>
<p>In order to prune the simulated data to approximate real data, RAREsim estimates the number of variants within MAC bins. This is either done by using default parametes, modifying default parameters or by fitting target data. Additionally, if wanting to simulate data the exact sample size of observed data, the observed data can be matched directly. Here we will demonstrate fitting target data as well as using default parameters.</p>
</div>
<div id="fitting-the-variants-per-kb-function" class="section level2">
<h2>Fitting the Variants per Kb function</h2>
<p>The first function is the Variants per Kb function. For a given region, the Variants per Kb function estimates the number of variants per Kb. This is done by estimating <span class="math inline">\(\phi\)</span> and <span class="math inline">\(\omega\)</span> to optimize the function <span class="math inline">\(f_{Var}(n)=\phi n^\omega\)</span>.</p>
<p>The Variants per Kb target data consists of various sample sizes and the observed number of variants per Kb in the region of interest. Ancestry specific data is advised.</p>
<p>The first column is the sample size is number of individuals and the second column is the observed number of variants per Kb in the region of interest. Here we will fit the example data for the African ancestry population, with example data available in the R package.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># load the data</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw">data</span>(<span class="st">&quot;var_per_kb_afr&quot;</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw">print</span>(var_per_kb_afr)</a></code></pre></div>
<pre><code>##    downsample     per_kb
## 1          10  0.2627568
## 2          20  0.6831678
## 3          50  1.5239897
## 4         100  2.7326712
## 5         200  4.3092123
## 6         500  7.6199485
## 7        1000 12.1919176
## 8        2000 19.3914551
## 9        3070 25.2246571
## 10       5000 33.4226707
## 11       5040 33.7905302
## 12       8128 45.1941773</code></pre>
<p>The target data is used to estimate <span class="math inline">\(\phi\)</span> and <span class="math inline">\(\omega\)</span> within a least squares loss function, optimizing using sequential quadratic programming. It is implemented with the <em>Fit_var_per_kb</em> function.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">Fit_var_per_kb</span>(var_per_kb_afr)</a></code></pre></div>
<pre><code>## $phi
## [1] 0.1638108
## 
## $omega
## [1] 0.6248848</code></pre>
<p>The output of the <em>Fit_total_variants</em> function are the parameters <span class="math inline">\(\phi\)</span> and <span class="math inline">\(\omega\)</span>, respectively. The estimated parameters can then be used to estimate the number of variants per Kb within the region of interest, given the number of individuals to be simulated <span class="math inline">\(N_{sim}\)</span>.</p>
<p>For example, to simulate the sample size observed in gnomAD for the African ancestry group (<span class="math inline">\(N_{sim}=8128\)</span>), we calculate <span class="math inline">\(f_{Var}(N_{sim})=\phi N_{sim}^\omega\)</span>. This can be done with the <em>Total_number_variants</em> function. A value for phi, omega, and the sample size is required.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">Variants_per_Kb</span>(<span class="dt">phi =</span> <span class="fl">0.1638108</span>, <span class="dt">omega =</span> <span class="fl">0.6248848</span>, <span class="dt">n =</span> <span class="dv">8128</span>)</a></code></pre></div>
<pre><code>## [1] 45.46026</code></pre>
<p>The number of variants per Kb was estimated using parameters fit with the target data. However, RAREsim also provides ancestry specific default parameters that can be used to estimate the number of variants per Kb.</p>
<p>Here, the ancestry must be specified: AFR, EAS, NFE, or SAS.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">Variants_per_Kb</span>(<span class="dt">n=</span><span class="dv">8128</span>, <span class="dt">pop =</span> <span class="st">'AFR'</span>) </a></code></pre></div>
<pre><code>## [1] 43.66395</code></pre>
<p>The cM block we are simulating has a total of 19,029 bp. Thus, to get the total expected number of variants in the region, we multiple the expected number of variants per Kb by 19.029.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="fl">19.029</span><span class="op">*</span><span class="kw">Variants_per_Kb</span>(<span class="dt">phi =</span> <span class="fl">0.1638108</span>, <span class="dt">omega =</span> <span class="fl">0.6248848</span>, <span class="dt">n =</span> <span class="dv">8128</span>)</a></code></pre></div>
<pre><code>## [1] 865.0633</code></pre>
<p>At this point, we know how many total variants we expect to see within the region. We now need the allele frequency spectrum (AFS) to calculate the proportion of variants at various minor allele bins.</p>
</div>
<div id="fitting-the-allele-frequency-spectrum-function" class="section level2">
<h2>Fitting the Allele Frequency Spectrum function</h2>
<p>The AFS function inputs allele frequency information and outputs estimates for <span class="math inline">\(\alpha\)</span>, <span class="math inline">\(\beta\)</span>, and <span class="math inline">\(b\)</span>. The required AFS target data is a vector with the number of variants in each MAC bin. The bins used within the evaluation of RAREsim are:</p>
<p>MAC=1<br />
MAC=2<br />
MAC=3-5<br />
MAC=6-10<br />
MAC=11-20<br />
MAC=21 - MAF = 0.5%<br />
MAC = 0.5% - MAF = 1%</p>
<p>Below we examine in the AFS target data for the African ancestry group. The first two columns identify the upper and lower boundaries of each MAC bin. The third column specifies the observed proportion of variants within each MAC bin in the target data.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># load the data</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="kw">data</span>(<span class="st">&quot;afs_afr&quot;</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="kw">print</span>(afs_afr)</a></code></pre></div>
<pre><code>##   Lower Upper       prop
## 1     1     1 0.50257998
## 2     2     2 0.16305470
## 3     3     5 0.08255934
## 4     6    10 0.05882353
## 5    11    20 0.03715170
## 6    21    81 0.05675955
## 7    82   162 0.01754386</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">Fit_AFS</span>(<span class="dt">prop_df =</span> afs_afr, <span class="dt">N =</span> <span class="dv">8128</span>, <span class="dt">p_rv =</span> <span class="fl">0.97</span>)</a></code></pre></div>
<pre><code>## $alpha
## [1] 1.531338
## 
## $beta
## [1] -0.3090162
## 
## $b
## [1] 0.2926182
## 
## $Fitted_results
##   Lower Upper        Fitted_prop
## 1     1     1  0.515383183430996
## 2     2     2  0.130900952615951
## 3     3     5  0.131313223208255
## 4     6    10 0.0689151687234886
## 5    11    20 0.0488202660410061
## 6    21    81 0.0582841702824958
## 7    82   162 0.0163830356978067</code></pre>
</div>
<div id="expected-number-of-variants-per-mac-bin" class="section level2">
<h2>Expected Number of Variants per MAC bin</h2>
<p>Now that all the parameters have been fit, the expected number of variants in each MAC bin can be extimated</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">bin_estimates &lt;-<span class="st"> </span><span class="kw">Expected_variants</span>(<span class="dt">alpha =</span> <span class="fl">1.5313376</span> , <span class="dt">beta =</span> <span class="fl">-0.3090162</span>, <span class="dt">b =</span> <span class="fl">0.2926182</span>, <span class="dt">phi =</span> <span class="fl">0.1638108</span> , <span class="dt">omega =</span> <span class="fl">0.6248848</span> , <span class="dt">Ntar =</span> <span class="dv">8128</span>, <span class="dt">Size =</span> <span class="fl">19.029</span>)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="kw">print</span>(bin_estimates)</a></code></pre></div>
<pre><code>##   Lower Upper Expected_var
## 1     1     1    445.83905
## 2     2     2    113.23761
## 3     3     5    113.59426
## 4     6    10     59.61599
## 5    11    20     42.23262
## 6    21    82     50.71816
## 7    81   162     14.47670</code></pre>
<p>We can also extimate the number of variants per MAC using the default parameters, again specifying the ancestry population as AFR, EAS, NFE, or SAS.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">bin_estimates &lt;-<span class="st"> </span><span class="kw">Expected_variants</span>(<span class="dt">pop=</span><span class="st">'AFR'</span> , <span class="dt">Ntar =</span> <span class="dv">8128</span>, <span class="dt">Size =</span> <span class="fl">19.029</span>)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="kw">print</span>(bin_estimates)</a></code></pre></div>
<pre><code>##   Lower Upper Expected_var
## 1     1     1    428.53078
## 2     2     2    103.53268
## 3     3     5     99.97549
## 4     6    10     50.23531
## 5    11    20     34.24618
## 6    21    82     38.90993
## 7    81   162     10.43360</code></pre>
<p>XXX right now the bins are hard coded in. Need to make them flexible to fit the rest of the functions XXX</p>
</div>
<div id="pruning-variants" class="section level2">
<h2>Pruning Variants</h2>
<p>In order to use RAREsim to prune simulated data, genetic data must be simulated with HAPGEN2, with all bp added to the input haplotypes. This will simulate an abudance of rare variants to allow for variant pruning.</p>
<p>Pruning variants required a MAC file fo the simulated data, and the expected number of variants within each MAC bin (what is produced fom the <em>Expected_variants</em> function).</p>
<p>The pruning function inputs the name of the gzipped haplotype file, a MAC file from the haplotypes (i.e. number of alternate alleles at each variant), and the expected number of variants per bin data frame, with the bins defined.</p>
<p>Having a MAC file allows the pruning process to be really fast. When there aren’t any alleles that need to move bins.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">data</span>(<span class="st">&quot;MAC_afr&quot;</span>)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw">dim</span>(MAC_afr)</a></code></pre></div>
<pre><code>## [1] 19029     1</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw">head</span>(MAC_afr)</a></code></pre></div>
<pre><code>##   V1
## 1  0
## 2  2
## 3  0
## 4  0
## 5  0
## 6  0</code></pre>
<p>When running the pruning function the haplotypes files will be re-written. Thus, if the original files are needed, a copy needs to be made.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"> <span class="kw">Pruning_function</span>(<span class="dt">hap_file_name =</span> <span class="st">'/Users/megansorenson/Documents/RAREsim/Example/Block37_rep1.controls.haps.gz'</span>, <span class="dt">MAC =</span> MAC_afr, <span class="dt">expected =</span> bin_estimates, <span class="dt">mac =</span> <span class="st">'mac'</span>)</a></code></pre></div>
<pre><code>## [1] 2616</code></pre>
<p>The output in R simply states tthe number of variants that were pruned. However, the haplotype files were changed.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
